src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -772,4 +784,183 @@ public void delete(Long programmingExerciseId, boolean deleteBaseReposBuildPlans
+        final var templatePlanName = BuildPlanType.TEMPLATE.getName();
+        final var solutionPlanName = BuildPlanType.SOLUTION.getName();
+        final var templateKey = templateExercise.getProjectKey();
+        final var targetKey = newExercise.getProjectKey();
+        final var targetName = newExercise.getCourse().getShortName().toUpperCase() + " " + newExercise.getTitle();
+        final var targetExerciseProjectKey = newExercise.getProjectKey();
+
+        // Clone all build plans, enable them and setup the initial participations, i.e. setting the correct rep URLs and
+        // running the plan for the first time
+        continuousIntegrationService.get().copyBuildPlan(templateKey, templatePlanName, targetKey, targetName, templatePlanName);
+        continuousIntegrationService.get().copyBuildPlan(templateKey, solutionPlanName, targetKey, targetName, solutionPlanName);
+        continuousIntegrationService.get().enablePlan(templateParticipation.getBuildPlanId());
+        continuousIntegrationService.get().enablePlan(solutionParticipation.getBuildPlanId());
+        continuousIntegrationService.get().updatePlanRepository(targetExerciseProjectKey, templateParticipation.getBuildPlanId(), ASSIGNMENT_REPO_NAME, targetExerciseProjectKey,
+                newExercise.getTemplateRepositoryName());
+        continuousIntegrationService.get().updatePlanRepository(targetExerciseProjectKey, solutionParticipation.getBuildPlanId(), ASSIGNMENT_REPO_NAME, targetExerciseProjectKey,
+                newExercise.getSolutionRepositoryName());
+        try {
+            continuousIntegrationService.get().triggerBuild(templateParticipation);
+            continuousIntegrationService.get().triggerBuild(solutionParticipation);
+        }
+        catch (HttpException e) {
+            log.error("Unable to trigger imported build plans", e);
+            throw e;
+        }
+    }
+
+    /**
+     * Copied test cases from one exercise to another. The test cases will get new IDs, thus being saved as a new entity.
+     * The remaining contents stay the same, especially the weights.
+     *
+     * @param templateExercise The template exercise which test cases should get copied
+     * @param targetExercise The new exercise to which all test cases should get copied to
+     */
+    private void importTestCases(final ProgrammingExercise templateExercise, final ProgrammingExercise targetExercise) {
+        targetExercise.setTestCases(templateExercise.getTestCases().stream().map(testCase -> {
+            final var copy = new ProgrammingExerciseTestCase();
+
+            // Copy everything except for the referenced exercise
+            copy.setActive(testCase.isActive());
+            copy.setAfterDueDate(testCase.isAfterDueDate());
+            copy.setTestName(testCase.getTestName());
+            copy.setWeight(testCase.getWeight());
+            copy.setExercise(targetExercise);
+            return copy;
+        }).collect(Collectors.toSet()));
```

Comment:
I guess the problem identified by Thilo is caused here. You do not save the test case entity into the database...

Reply:
I now save the entity properly